x-pack/plugins/security_solution/public/common/components/guided_onboarding_tour/tour.tsx

```diff
@@ -0,0 +1,111 @@
+
+import useObservable from 'react-use/lib/useObservable';
+import { catchError, of, timeout } from 'rxjs';
+import { useKibana } from '../../lib/kibana';
+import { securityTourConfig, SecurityStepId } from './tour_config';
+
+export interface TourContextValue {
+  activeStep: number;
+  endTourStep: (stepId: SecurityStepId) => void;
+  incrementStep: (stepId: SecurityStepId, step?: number) => void;
+  isTourShown: (stepId: SecurityStepId) => boolean;
+}
+
+const initialState: TourContextValue = {
+  activeStep: 0,
+  endTourStep: () => {},
+  incrementStep: () => {},
+  isTourShown: () => false,
+};
+
+const TourContext = createContext<TourContextValue>(initialState);
+
+export const TourContextProvider = ({ children }: { children: ReactChild }) => {
+  const { guidedOnboardingApi } = useKibana().services.guidedOnboarding;
+
+  const isRulesTourActive = useObservable(
+    guidedOnboardingApi?.isGuideStepActive$('security', SecurityStepId.rules).pipe(
+      // if no result after 30s the observable will error, but the error handler will just emit false
+      timeout(30000),
+      catchError((error) => of(false))
+    ) ?? of(false),
+    false
+  );
+  const isAlertsCasesTourActive = useObservable(
+    guidedOnboardingApi?.isGuideStepActive$('security', SecurityStepId.alertsCases).pipe(
+      // if no result after 30s the observable will error, but the error handler will just emit false
+      timeout(30000),
+      catchError((error) => of(false))
+    ) ?? of(false),
+    false
+  );
+
+  const tourStatus = useMemo(
+    () => ({
+      [SecurityStepId.rules]: isRulesTourActive,
+      [SecurityStepId.alertsCases]: isAlertsCasesTourActive,
+    }),
+    [isRulesTourActive, isAlertsCasesTourActive]
+  );
+
+  const isTourShown = useCallback((stepId: SecurityStepId) => tourStatus[stepId], [tourStatus]);
+  const [activeStep, _setActiveStep] = useState<number>(1);
+
+  const incrementStep = useCallback((stepId: SecurityStepId, step?: number) => {
+    _setActiveStep((prevState) =>
+      step != null && step <= securityTourConfig[stepId].length
+        ? step
+        : (prevState >= securityTourConfig[stepId].length ? 0 : prevState) + 1
+    );
+  }, []);
+
+  const resetStep = useCallback(() => {
+    _setActiveStep(1);
+  }, []);
+
+  useEffect(() => {
+    resetStep();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [tourStatus]);
+
+  // TODO: @Steph figure out if we're allowing user to skip tour or not, implement this if so
+  // const onSkipTour = useCallback((stepId: SecurityStepId) => {
+  //   // active state means the user is on this step but has not yet begun. so when the user hits skip,
+  //   // the tour will go back to this step until they "re-start it"
+  //   // guidedOnboardingApi.idkSetStepTo(stepId, 'active')
+  // }, []);
+
+  const endTourStep = useCallback(
+    async (stepId: SecurityStepId) => {
+      await guidedOnboardingApi?.completeGuideStep('security', stepId);
+      resetStep();
```

Comment:
If the component is unmounted while the `await ...` is executing the `resetStep()` will throw the following error:
```
Warning: Can't perform a React state update on an unmounted
component. This is a no-op, but it indicates a memory leak in
your application. To fix, cancel all subscriptions and
asynchronous tasks in a useEffect cleanup function.
```
This error will only rise locally, but it would be nice to take care of these situations when managing side effects.

Reply:
thank you, will investigate

Reply:
I was not able to reproduce this error myself, but I made a commit that adds a mount check to prevent this: https://github.com/elastic/kibana/pull/143598/commits/e09574aea2ba2ebdd0c0a35e721b34bb99cd1baa